### PR TITLE
Support JobSet metrics

### LINKED ISSUE #2560: Support for JobSet metrics
**What would you like to be added**:

Expose metrics about JobSet resource: https://jobset.sigs.k8s.io/docs/reference/jobset.v1alpha2/

**Why is this needed**:

For better monitoring of AI training job status.

**Describe the solution you'd like**

Following docs/developer/guide.md to add a new JobSet resource with a few metrics reporting the status of the JobSet.

**Additional context**

### PR DESCRIPTION
**Pull request by chelseychen**
**What this PR does / why we need it**:

Add support for JobSet metrics. Make JobSet resource optional.

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*
The change won't affect cardinality of KSM by default since JobSet is added as an optional resource and will not be enabled by default.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2560

### CODE REVIEW

File: examples/standard/cluster-role.yaml [hunk 1]
```diff
@@ -126,3 +126,10 @@ rules:
   verbs:
   - list
   - watch
+- apiGroups:
+  - jobset.x-k8s.io
```
Comment (mrueg) [inline examples/standard/cluster-role.yaml:130]: That looks like an external CRD that is not part of a default kubernetes install, is it?

Comment (chelseychen) [reply examples/standard/cluster-role.yaml:130]: Correct, this is still a CRD at the moment, but the demand for having jobset metrics is increasing and people are working on making it a native k8s api, so I'm adding it as a build-in resource here.

**Comment by chelseychen** [comment]
Tested in a cluster with 3 jobsets (from [examples](https://github.com/kubernetes-sigs/jobset/tree/main/site/static/examples/simple)) deployed, and see metrics showing up:
```
# HELP kube_jobset_annotations Kubernetes annotations converted to Prometheus labels.
# TYPE kube_jobset_annotations gauge
# HELP kube_jobset_labels Kubernetes labels converted to Prometheus labels.
# TYPE kube_jobset_labels gauge
# HELP kube_jobset_specified_replicas The Number of specified replicas per replicated jobs in a jobset.
# TYPE kube_jobset_specified_replicas gauge
kube_jobset_specified_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="leader"} 1
kube_jobset_specified_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="workers"} 1
kube_jobset_specified_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="leader"} 1
kube_jobset_specified_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="workers"} 1
kube_jobset_specified_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="leader"} 1
kube_jobset_specified_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="workers"} 1
# HELP kube_jobset_status_replicas The Number of replicas in ready/succeeded/failed/active/suspended status per replicated jobs in a jobset.
# TYPE kube_jobset_status_replicas gauge
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="workers",status="ready"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="workers",status="succeeded"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="workers",status="failed"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="workers",status="active"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="workers",status="suspended"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="leader",status="ready"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="leader",status="succeeded"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="leader",status="failed"} 1
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="leader",status="active"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="failure-policy",replicated_job_name="leader",status="suspended"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="workers",status="ready"} 1
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="workers",status="succeeded"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="workers",status="failed"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="workers",status="active"} 1
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="workers",status="suspended"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="leader",status="ready"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="leader",status="succeeded"} 1
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="leader",status="failed"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="leader",status="active"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="network-jobset",replicated_job_name="leader",status="suspended"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="workers",status="ready"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="workers",status="succeeded"} 1
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="workers",status="failed"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="workers",status="active"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="workers",status="suspended"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="leader",status="ready"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="leader",status="succeeded"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="leader",status="failed"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="leader",status="active"} 0
kube_jobset_status_replicas{namespace="default",jobset_name="success-policy",replicated_job_name="leader",status="suspended"} 0
# HELP kube_jobset_status_condition The current status conditions of a jobset.
# TYPE kube_jobset_status_condition gauge
kube_jobset_status_condition{namespace="default",jobset_name="failure-policy",condition="Failed",status="true"} 1
kube_jobset_status_condition{namespace="default",jobset_name="failure-policy",condition="Failed",status="false"} 0
kube_jobset_status_condition{namespace="default",jobset_name="failure-policy",condition="Failed",status="unknown"} 0
kube_jobset_status_condition{namespace="default",jobset_name="success-policy",condition="Completed",status="true"} 1
kube_jobset_status_condition{namespace="default",jobset_name="success-policy",condition="Completed",status="false"} 0
kube_jobset_status_condition{namespace="default",jobset_name="success-policy",condition="Completed",status="unknown"} 0
```

**Comment by mrueg** [comment]
Following up from https://github.com/kubernetes/kube-state-metrics/pull/2563/files#r1861201722

Supporting non-native resources as a built-in resource is not something that is in scope for ksm. This is also why we removed VPA metrics and likely https://github.com/kubernetes/kube-state-metrics/pull/2452 will never make it in.

We have 
https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md and (soon)
https://github.com/rexagod/resource-state-metrics to support non-native resources.

I think at the moment for JobSet the CRS metric config would be the best option. Whenever it becomes a native resource that is bundled with kubernetes, we're happy to include it.

**Comment by chelseychen** [comment]
> Following up from https://github.com/kubernetes/kube-state-metrics/pull/2563/files#r1861201722
> 
> Supporting non-native resources as a built-in resource is not something that is in scope for ksm. This is also why we removed VPA metrics and likely #2452 will never make it in.
> 
> We have https://github.com/kubernetes/kube-state-metrics/blob/main/docs/metrics/extend/customresourcestate-metrics.md and (soon) https://github.com/rexagod/resource-state-metrics to support non-native resources.
> 
> I think at the moment for JobSet the CRS metric config would be the best option. Whenever it becomes a native resource that is bundled with kubernetes, we're happy to include it.

I tried CRS config but it turns out the memory usage is higher and also it has some restrictions on metrics definitions (e.g., I found no way to define `kube_jobset_status_replicas` metrics with status field set to specific string values).

The purpose of adding JobSet as a build-in resource is to make the metrics definitions consistent across users. And it is optional not default-on so it only affects those who want to explicitly enable JobSet resource in KSM.

**Comment by mrueg** [comment]
Closing this as we won't support non-native resources.

**Pull request closed without merging**